x-pack/plugins/endpoint/server/routes/resolver/utils/pagination.ts

```diff
@@ -4,104 +4,159 @@
 
-export interface PaginationParams {
-  size: number;
-  timestamp?: number;
-  eventID?: string;
-}
-
-export interface PaginatedResults {
-  totals: Record<string, number>;
-  results: ResolverEvent[];
-  // content holder for any other extra aggregation counts
-  extras?: Record<string, Record<string, number>>;
+interface TotalsAggregation {
+  totals?: {
+    buckets?: AggBucket[];
+  };
 }
 
 interface PaginationCursor {
   timestamp: number;
   eventID: string;
 }
 
-function urlEncodeCursor(data: PaginationCursor): string {
-  const value = JSON.stringify(data);
-  return Buffer.from(value, 'utf8')
-    .toString('base64')
-    .replace(/\+/g, '-')
-    .replace(/\//g, '_')
-    .replace(/=+$/g, '');
+/**
+ * The result structure of a query that leverages pagination. This includes totals that can be used to determine if
+ * additional nodes exist and additional queries need to be made to retrieve the nodes.
+ */
+export interface PaginatedResults {
+  /**
+   * Resulting events returned from the query.
+   */
+  results: ResolverEvent[];
+  /**
+   * Mapping of unique ID to total number of events that exist in ES. The events this references is scoped to the events
+   * that the query is searching for.
+   */
+  totals: Record<string, number>;
 }
 
-function urlDecodeCursor(value: string): PaginationCursor {
-  value = value.replace(/\-/g, '+').replace(/_/g, '/');
-  const data = Buffer.from(value, 'base64').toString('utf8');
-  const { timestamp, eventID } = JSON.parse(data);
-  // take some extra care to only grab the things we want
-  // convert the timestamp string to date object
-  return { timestamp, eventID };
-}
+/**
+ * This class handles constructing pagination cursors that resolver can use to return additional events in subsequent
+ * queries. It also constructs an aggregation query to determine the totals for other queries. This class should be used
+ * with a query to build cursors for paginated results.
+ */
+export class PaginationBuilder {
+  constructor(
+    /**
+     * upper limit of how many results should be returned by the parent query.
+     */
+    private readonly size: number,
+    /**
+     * timestamp that will be used in the search_after section
+     */
+    private readonly timestamp?: number,
+    /**
+     * unique ID for the last event
+     */
+    private readonly eventID?: string
+  ) {}
 
-export function getPaginationParams(limit: number, after?: string): PaginationParams {
-  if (after) {
-    try {
-      const cursor = urlDecodeCursor(after);
-      if (cursor.timestamp && cursor.eventID) {
-        return {
-          size: limit,
-          timestamp: cursor.timestamp,
-          eventID: cursor.eventID,
-        };
-      }
-    } catch (err) {
-      /* tslint:disable:no-empty */
-    } // ignore invalid cursor values
+  private static urlEncodeCursor(data: PaginationCursor): string {
+    const value = JSON.stringify(data);
+    return Buffer.from(value, 'utf8')
+      .toString('base64')
```

Comment:
:grey_question: Random thought: `encodeURI( new StringDecoder('base64').write(Buffer.from(value, 'utf8')) )`

Reply:
nvm, I was thinking of encodeURIComponent ^ that won't work. Carry on :rofl: